### PR TITLE
osd/PG: fix history.same_interval_since of merge target again

### PR DESCRIPTION
**Pull request by xiexingguo**
The symptom looks much like we see in
	https://tracker.ceph.com/issues/37654.

The root cause is that both merge source and target could be
fabricated PGs (aka placeholders), hence merge target's
same_primary_since could remain 0 after merge.

Fix by adjusting history.same_primary_since to last_epoch_clean
reported by these PGs were found to be ready for merge.
This peer is going to be ignored/purged by primary anyway later
when peering is done.

Fixes: https://tracker.ceph.com/issues/45991
Signed-off-by: xie xingguo <xie.xingguo@zte.com.cn>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by xiexingguo** [comment]
> did you mean "history.same_interval_since" in your commit message?

yup :)

File: src/osd/PeeringState.cc [hunk 1]
```diff
@@ -3086,6 +3086,13 @@ void PeeringState::merge_from(
 	       << sources.begin()->second->info.history
 	       << dendl;
 
+    if (info.history.epoch_created == 0) {
```
Comment (neha-ojha) [inline src/osd/PeeringState.cc:3089]: We already check for this before getting here, see https://github.com/ceph/ceph/blob/master/src/osd/PeeringState.cc#L3071. Also, how about doing this adjustment for info.history.same_interval_since with https://github.com/ceph/ceph/blob/master/src/osd/PeeringState.cc#L3110-L3121, where the existing adjustment is done?

Comment (xiexingguo) [reply src/osd/PeeringState.cc:3089]: > We already check for this before getting here, see https://github.com/ceph/ceph/blob/master/src/osd/PeeringState.cc#L3071.

That line is for merge target. If we hit that condition, merge target should be either a placeholder or a newly initiated peer that never got the chance to finish a complete peering cycle, then we'll try to steal info.history from merge source(we do that exact two lines below, see https://github.com/ceph/ceph/blob/master/src/osd/PeeringState.cc#L3074).
But the merge source could be a placeholder too, that's why we need to fix history.sis again.



> Also, how about doing this adjustment for info.history.same_interval_since with https://github.com/ceph/ceph/blob/master/src/osd/PeeringState.cc#L3110-L3121, where the existing adjustment is done?

That adjustment should happen only if we already have a meaningful history.same_interval_since in hand.

Comment (neha-ojha) [reply src/osd/PeeringState.cc:3089]: makes sense, could you add this explanation as a comment for clarity

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2020-06-17_01:02:07-rados-wip-kefu-testing-2020-06-17-0030-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2020-06-17_08:41:50-rados-wip-kefu-testing-2020-06-17-1349-distro-basic-smithi/

test failure is tracked by https://tracker.ceph.com/issues/46046

**Comment by tchaikov** [comment]
@xiexingguo shall we backport this change to octopus and probably nautilus?

**Comment by xiexingguo** [comment]
> @xiexingguo shall we backport this change to octopus and probably nautilus?

both octopus and nautilus.

@tchaikov Thanks for testing!

**Pull request merged**